libraries/standard/json/src/json.c

```diff
@@ -0,0 +1,1014 @@
+    size_t i = *start;
+
+    while( i < max )
+    {
+        c = buf[ i ];
+        i++;
+
+        switch( c )
+        {
+            case '{':
+            case '[':
+                depth++;
+
+                if( depth == JSON_MAX_DEPTH )
+                {
+                    ret = JSONMaxDepthExceeded;
+                    break;
+                }
+
+                stack[ depth ] = c;
+                break;
+
+            case '}':
+            case ']':
+
+                if( depth > 0 )
+                {
+                    depth--;
+                    ( void ) skipSpaceAndComma( buf, &i, max );
+                    break;
+                }
+
+                ret = ( depth == 0 ) ? JSONSuccess : JSONIllegalDocument;
+                break;
+
+            default:
+                ret = JSONIllegalDocument;
+                break;
+        }
+
+        if( ret != JSONPartial )
+        {
+            break;
+        }
+
+        ( void ) skipScalars( buf, &i, max, stack[ depth ] );
+    }
+
+    *start = i;
+    return ret;
+}
+
+/**
+ * See json.h for docs.
+ *
+ * Verify that the entire buffer contains exactly one scalar
+ * or collection within optional whitespace.
+ */
+JSONStatus_t JSON_Validate( const char * buf,
+                            size_t length )
+{
+    JSONStatus_t ret;
+    size_t i = 0U;
+
+    assert( ( buf != NULL ) && ( length > 0U ) );
+
+    skipSpace( buf, &i, length );
+    #ifndef JSON_VALIDATE_COLLECTIONS_ONLY
+        if( skipAnyScalar( buf, &i, length ) == true )
+        {
+            ret = JSONSuccess;
+        }
+        else
+    #endif
+    {
+        ret = match( buf, &i, length );
+    }
+
+    if( ( ret == JSONSuccess ) && ( i < length ) )
+    {
+        skipSpace( buf, &i, length );
+
+        if( i != length )
+        {
+            ret = JSONIllegalDocument;
+        }
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Output indexes for the next key-value pair of an object.
+ *
+ * Also advances the buffer index beyond the key-value pair.
+ * The value may be a scalar or a collection.
+ *
+ * @param[in] buf  The buffer to parse.
+ * @param[in,out] start  The index at which to begin.
+ * @param[in] max  The size of the buffer.
+ * @param[out] key  A pointer to receive the index of the key.
+ * @param[out] keyLength  A pointer to receive the length of the key.
+ * @param[out] value  A pointer to receive the index of the value.
+ * @param[out] valueLength  A pointer to receive the length of the value.
+ *
+ * @return true if a key-value pair was present;
+ * false otherwise.
+ */
+static bool nextKeyValuePair( const char * buf,
+                              size_t * start,
+                              size_t max,
+                              size_t * key,
+                              size_t * keyLength,
+                              size_t * value,
+                              size_t * valueLength )
+{
+    bool ret = true;
+    size_t i = *start, save = *start;
```

Comment:
I think it would be easier to read this function if you split "save" into two separate variables called something like "keyStart" and "valueStart". I got a little confused while reading because I didn't initially guess what the name "save" meant, and I also didn't notice that it was used twice.

Reply:
changed